dbms/src/IO/Progress.h

```diff
@@ -6,26 +6,71 @@
 
 #include <Core/Defines.h>
 
-
 namespace DB
 {
 
 class ReadBuffer;
 class WriteBuffer;
 
+struct AllProgressValueImpl;
 
 /// See Progress.
 struct ProgressValues
 {
     size_t rows;
     size_t bytes;
     size_t total_rows;
+    size_t write_rows;
+    size_t write_bytes;
 
-    void read(ReadBuffer & in, UInt64 server_revision);
-    void write(WriteBuffer & out, UInt64 client_revision) const;
-    void writeJSON(WriteBuffer & out) const;
+
+    template <typename ReadImpl = AllProgressValueImpl>
+    inline void read(ReadBuffer & in, UInt64 server_revision);
+
+    template <typename WriteImpl = AllProgressValueImpl>
+    inline void write(WriteBuffer & out, UInt64 client_revision) const;
+
+    template <typename WriteJSONImpl = AllProgressValueImpl>
+    inline void writeJSON(WriteBuffer & out) const;
 };
 
+struct AllProgressValueImpl
```

Comment:
Lack of comments. (but first need to think how to rewrite this code...)

Reply:
Let's both ReadProgress and WriteProgress have read, write, writeJSON methods.
But better to rename read -> deserializeBinary, write -> serializeBinary, writeJSON -> serializeJSON for clarity.
And let's rename AllProgress to ReadWriteProgress.
And let's ReadWriteProgress has ReadProgress and WriteProgress as its members. And its serde methods will take revision as an argument and call child methods appropriately.